### ISSUE TITLE
Any plan to support flash attention 3 for hopper GPUs?

### ISSUE THREAD

**Issue by fno2010** [pytorch/pytorch#137901]
### 🚀 The feature, motivation and pitch

Flash Attention 3 (https://github.com/Dao-AILab/flash-attention) has been in beta for some time. I tested it on H100 GPUs with CUDA 12.3 and also attempted a simple merge with PyTorch. I understand that it will take some time before the stable release is ready, but I believe some structural refactoring can be started first. e.g., The FA3 highly relies on the hopper architecture. Therefore, for other architectures, the `enable_flash` option of the SDP kernel still needs to go to FA2 backend.

I wonder if anyone has already planned or been working on this. As I can see, @drisspg is still working on adding 2.6.2. I am willing to contribute but don't want to repeat any effort.

**Comment by Skylion007** [comment]
We need an internal contributor to champion #131493 so we can update CUTLASS to a version that would support it.

**Comment by fno2010** [comment]
@Skylion007 Thanks for the clarification. I will follow up on the CUTLASS upgrade issue. In my local test, I separately compile `flash_attn` to a static library with CUTLASS 3.5.1 to avoid the upgrade issue. But this trick should not be used in the release.

About CUDNN 9.4, I will try more benchmarks. From our previous benchmark results, CUDNN performs similar or even better than FA3 in small sequence lengths, but FA3 still improves in large sequence lengths.

**Comment by drisspg** [comment]
You might have answered my exact question have you profiled the new CuDNN backend? It seems like the FAv3 codebase is still pretty rapidly changing but last I profiled the CuDNN backend was at the same performance if not a little better on the same shapes in the FAv3 paper for bf16 inputs. 

Do you have a repro for showing the large sequence length perf increase?

**Comment by fno2010** [comment]
I have run the forward attention benchmark for FAV3 and CuDNN {9.0, 9.1, 9.4, 9.5} in a single H100. Let me post the results:

![fav3_vs_cudnn_causal_hdim64](https://github.com/user-attachments/assets/f1b9cf5d-0cb2-4ceb-b3a5-7f2aa24847f6)
![fav3_vs_cudnn_causal_hdim128](https://github.com/user-attachments/assets/17dff9c1-dacb-4d0c-9aca-e5e5ed99ab0f)
![fav3_vs_cudnn_nocausal_hdim64](https://github.com/user-attachments/assets/196390ab-faa9-4b60-9e65-9b656ff67f5f)
![fav3_vs_cudnn_nocausal_hdim128](https://github.com/user-attachments/assets/b75482ea-d009-431f-985b-329fc40abc2f)

But maybe my FAV3 is a little bit old (based on `30e1ef0f7`, a commit one month ago). Also, I haven't tested the backward performance yet. I will follow up the latest FAV3 updates and get back to the pytorch community soon.

**Comment by danya0123** [comment]
any chance to support FAv3 for fp8 training?

**Comment by WingsOfPanda** [comment]
any update with fa3 and pytorch integration?

**Comment by jbschlosser** [comment]
I'm looking into this semi-actively (while juggling a couple other things). If we provide FA3 as an -additional- option beyond FA2, this should be doable more quickly than waiting for dropout support for BC.

**Comment by KevinZeng08** [comment]
Can SDPA call the FlashAttention3 with some configurations currently?

**Comment by jbschlosser** [comment]
@KevinZeng08 no not yet, FA3 is only available right now as a separate package outside of PyTorch.

**Comment by drisspg** [comment]
@KevinZeng08 In the interim you can manually enable CuDNN similar to https://github.com/pytorch/torchtitan/blob/49d5cab1c32ec5ae2bbe4fab47b4f69725083eec/torchtitan/models/attention.py#L183

we are planning to not make this an opt in feature (so that you can get the speed boost transparently) in a future release but for now you can manually specify it

**Comment by drisspg** [comment]
Closing as we have since landed support: https://docs.pytorch.org/docs/2.12/generated/torch.nn.attention.activate_flash_attention_impl.html#torch.nn.attention.activate_flash_attention_impl

We have abi compat wheels here: https://download.pytorch.org/whl/flash-attn-3/

FA4 wheels can be found here: https://pypi.org/project/flash-attn-4/
